[@kbn/config-schema] add `discriminatedUnion ` schema type

nickofthyme opened the pull request:
## Summary

Adds `discriminatedUnion` method to `@kbn/config-schema` for discriminated union of object schemas.

```ts
const exampleType = schema.discriminatedUnion('type', [
  schema.object({ type: schema.literal('str'), string: schema.string() }),
  schema.object({ type: schema.literal('num'), number: schema.number() }),
  schema.object({ type: schema.literal('bool'), boolean: schema.boolean() }),
]);
```

Order of Errors:
1) missing `discriminator` property.
1) `discriminator` property has wrong type (only strings via `schema.literal` and one `schema.string` are permitted)
1) `discriminator` is a string but not one of the expected discriminators.
1) validation errors from the respective schema based on `discriminator`, only the matching schema is validated.


Also allows for fallback schema. Limit of one fallback, position in array does not matter.

```ts
const exampleType = schema.discriminatedUnion('type', [
  schema.object({ type: schema.literal('str'), string: schema.string() }),
  schema.object({ type: schema.literal('num'), number: schema.number() }),
  schema.object({ type: schema.string }, { unknowns: 'allow' }), // allows any other type other than 'str' and 'num'
]);
```

## Error Example

```
{ type: 'str', string: 123 } // "string" property is a number instead of a string

[string]: expected value of type [string] but got [number]
```

closes #84264

### Checklist

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

nickofthyme committed:
feat: add `oneOfKind` schema type

nickofthyme committed:
chore: cleanup types

- remove `DiscriminatorType`, not really needed
- fix return type of `oneOfKind` to return correct resolved type
- add type tests

nickofthyme committed:
improve `oneOf` types

nickofthyme committed:
fix: `defaultValue` typings on `oneOfKind`

nickofthyme committed:
Merge branch 'main' into schema-one-of-kind

nickofthyme committed:
Revert "improve `oneOf` types"

This reverts commit 9f0525ddc1e47764c8b16e88da3c45758e55e28f.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/378103)
* Commit: 05ca1fabd98a6a669a73c866ffb6a80ab382bd03

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/a8ceb3de3ff5fd02b0d54e57fa798acc8750139c) | [after](https://github.com/elastic/kibana/commit/05ca1fabd98a6a669a73c866ffb6a80ab382bd03) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1395 | 1396 | +1 |
| `dashboard` | 996 | 997 | +1 |
| `discover` | 1950 | 1951 | +1 |
| `infra` | 1772 | 1773 | +1 |
| `lens` | 1661 | 1662 | +1 |
| `observability` | 1688 | 1689 | +1 |
| `observabilityAIAssistantApp` | 800 | 801 | +1 |
| `securitySolution` | 8591 | 8592 | +1 |
| `stackConnectors` | 637 | 638 | +1 |
| `unifiedDocViewer` | 778 | 779 | +1 |
| total |  |  | +10 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/a8ceb3de3ff5fd02b0d54e57fa798acc8750139c) | [after](https://github.com/elastic/kibana/commit/05ca1fabd98a6a669a73c866ffb6a80ab382bd03) | diff |
| --- | --- | --- | --- |
| `@kbn/config-schema` | 143 | 144 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/a8ceb3de3ff5fd02b0d54e57fa798acc8750139c) | [after](https://github.com/elastic/kibana/commit/05ca1fabd98a6a669a73c866ffb6a80ab382bd03) | diff |
| --- | --- | --- | --- |
| `lens` | 1.9MB | 1.9MB | +1.3KB |
| `observability` | 1.7MB | 1.7MB | +1.3KB |
| `securitySolution` | 10.7MB | 10.7MB | +1.3KB |
| `stackConnectors` | 1.1MB | 1.1MB | +1.3KB |
| total |  |  | +5.1KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/a8ceb3de3ff5fd02b0d54e57fa798acc8750139c) | [after](https://github.com/elastic/kibana/commit/05ca1fabd98a6a669a73c866ffb6a80ab382bd03) | diff |
| --- | --- | --- | --- |
| `@kbn/config-schema` | 20 | 22 | +2 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/a8ceb3de3ff5fd02b0d54e57fa798acc8750139c) | [after](https://github.com/elastic/kibana/commit/05ca1fabd98a6a669a73c866ffb6a80ab382bd03) | diff |
| --- | --- | --- | --- |
| `lens` | 64.6KB | 64.6KB | -2.0B |
| `stackConnectors` | 76.7KB | 76.7KB | +3.0B |
| total |  |  | +1.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/a8ceb3de3ff5fd02b0d54e57fa798acc8750139c) | [after](https://github.com/elastic/kibana/commit/05ca1fabd98a6a669a73c866ffb6a80ab382bd03) | diff |
| --- | --- | --- | --- |
| `@kbn/config-schema` | 150 | 151 | +1 |


</details>


### History
* :broken_heart: [Build #378063](https://buildkite.com/elastic/kibana-pull-request/builds/378063) failed 966a77222555ade15d8aae6c6e689cd6bbe4ad1c
* :broken_heart: [Build #372767](https://buildkite.com/elastic/kibana-pull-request/builds/372767) failed 4cb0f66a03b084689cdc6103ab11f2097035717e
* :yellow_heart: [Build #372474](https://buildkite.com/elastic/kibana-pull-request/builds/372474) was flaky d50e8e17f9bd3da5ea92be65cb6cfe1e944096e4

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/378103","number":378103,"commit":"05ca1fabd98a6a669a73c866ffb6a80ab382bd03"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/378063","number":378063,"commit":"966a77222555ade15d8aae6c6e689cd6bbe4ad1c"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372767","number":372767,"commit":"4cb0f66a03b084689cdc6103ab11f2097035717e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/372474","number":372474,"commit":"d50e8e17f9bd3da5ea92be65cb6cfe1e944096e4"}],"number":378103}
buildkite-pr-comment-->

jloleysens commented on src/platform/packages/shared/kbn-config-schema/index.ts:
```diff
@@ -224,6 +227,239 @@ function oneOf<RTS extends Array<Type<any>>>(
   return new UnionType(types, options);
 }
 
+function oneOfKind<
```
Nit: do you think we can work in the word `discriminated` to this name?

nickofthyme committed:
refactor: allow falback and add schema checks

- allow fallback or catch-all schema
- validate 0 or 1 fallback schema
- validate unique discriminators

nickofthyme committed:
chore: update method naming, add alias for `union`

jloleysens commented on src/platform/packages/shared/kbn-config-schema/src/types/discriminated_union_type.ts:
```diff
@@ -0,0 +1,102 @@
+  private readonly typeOptions?: UnionTypeOptions<T>;
+
+  constructor(discriminator: Discriminator, types: RTS, options?: UnionTypeOptions<T>) {
+    const schema = internals
```
Got it, the main piece is just that the info is there which it appears to be! In post-processing we should convert this output you shared into the OAS discriminator.

jloleysens reviewed (approved):
Great work @nickofthyme , this looks like a solid version of discriminated union with solid DX.

nickofthyme committed:
Merge branch 'main' into schema-one-of-kind

nickofthyme committed:
chore: add `discriminator` to schema `meta`
